daemon/lib/source/DobbyRunC.cpp

```diff
@@ -173,20 +173,129 @@ std::pair<pid_t, pid_t> DobbyRunC::create(const ContainerId &id,
+            AI_LOG_DEBUG("Cannot kill after timeout");
+            exited_pid = waitpid(worker_pid, &status, WNOHANG);
+        }
+        else
+        {
+            // Worker is stuck, we need to kill it
+            AI_LOG_DEBUG("Can kill after timeout");
+            kill(worker_pid, SIGKILL);
+            // Collect the worker process
+            waitpid(worker_pid, &status, 0);
+            // Collect child of worker if any
+            wait(nullptr); 
+        }
+
+    }
+    else if (exited_pid == worker_pid)
+    {
+        // Worker finished
+        kill(timeout_pid, SIGKILL);
+        // Collect the timeout process
+        wait(nullptr); 
+    }
+
+
+    // Now as we had finished both forks we can safely exit if necessary
+    if (exited_pid == timeout_pid)
+    {
+        AI_LOG_WARN("Timeout occurred");
+
+        // We need to clean up after failed container creation, as we
+        // don't know when in creation process it failed do full step
+        // by step procedure
+
+        // First kill container so it is in stopped state
+        if (!killCont(id, SIGKILL))
+        {
+            // Even though the container couldn't be killed it can still
+            // exists so we still need to destroy it so no return here
+            AI_LOG_WARN("failed to kill (non-running) container for '%s'",
+                        id.c_str());
+        }
+        else
+        {
+            // We are not sure if process started, so check if we can get
+            // container pid, read the file
+            const size_t maxLength = 64;
+            std::string pidFileContents = mUtilities->readTextFile(pidFilePath, maxLength);
+            if (!pidFileContents.empty())
+            {
+                // and try to convert the file content into pid
+                char *endptr;
+                pid_t containerPid = static_cast<pid_t>(strtol(pidFileContents.c_str(), &endptr, 0));
+                if (endptr != pidFileContents.c_str())
+                {
+                    // wait for the half-started container to terminate
+                    if (waitpid(containerPid, nullptr, 0) < 0)
+                    {
+                        AI_LOG_SYS_ERROR(errno, "error waiting for (non-running) container '%s' to terminate",
+                                        id.c_str());
+                    }
+                }
+            }
+        }
+
+        // Don't bother capturing hook logs here
+        std::shared_ptr<DobbyDevNullStream> nullstream = std::make_shared<DobbyDevNullStream>();
+        AI_LOG_INFO("attempting to destroy (non-running) container '%s'", id.c_str());
+        // Force delete by default as we have no idea what condition the container is in
+        // and it may not respond to a normal delete
+        if (!destroy(id, nullstream, true))
+        {
+            AI_LOG_ERROR("failed to destroy '%s'", id.c_str());
+        }
+
+        return {-1,-1};
+    }
+    else if (exited_pid < 0)
```

Comment:
Can we invert this if statement to reduce nesting? (just for readability)

Make this the main `if`, then since we return here the other half doesn't need to be nested inside an else block

Reply:
This is to keep the logic same between other runc call methods. Also in each of them we are returning with {-1,-1} so we could get rid of else from everywhere, but as I said this is how every other method looks now.

Reply:
That's fine - consistency trumps neatness here